lib/phoenix_live_view/lifecycle.ex

```diff
@@ -0,0 +1,201 @@
+    invalid lifecycle event provided to attach_hook.
+
+    Expected one of: :handle_event | :handle_info | :handle_params
+
+    Got: #{inspect(stage)}
+    """
+  end
+
+  @doc false
+  def __attach_at_mount__(%__MODULE__{} = lifecycle, at_mount) when is_list(at_mount) do
+    Enum.reduce(at_mount, lifecycle, fn id, acc ->
+      {mod, fun} =
+        case id do
+          {mod, fun} when is_atom(mod) and is_atom(fun) -> {mod, fun}
+          mod when is_atom(mod) -> {mod, :mount}
+        end
+
+      hook = Hook.new!(id, :mount, Function.capture(mod, fun, 3))
+      %{acc | mount: [hook | acc.mount]}
+    end)
+  end
+
+  @doc """
+  TODO
+
+  Must match the id given to attach_hook/4
+  """
+  def detach_hook(%Socket{} = socket, id, stage)
+      when stage in [:handle_event, :handle_info, :handle_params] do
+    update_lifecycle(socket, stage, fn hooks ->
+      for hook <- hooks, hook.id != id, do: hook
+    end)
+  end
+
+  def detach_hook(%Socket{}, _id, stage) do
+    raise ArgumentError, """
+    invalid lifecycle event provided to detach_hook.
+
+    Expected one of: :handle_event | :handle_info | :handle_params
+
+    Got: #{inspect(stage)}
+    """
+  end
+
+  defp lifecycle(socket) do
+    Map.fetch!(socket.private, @lifecycle)
+  end
+
+  defp update_lifecycle(socket, stage, fun) do
+    lifecycle = lifecycle(socket)
+    new_lifecycle = Map.update!(lifecycle, stage, fun)
+    put_lifecycle(socket, new_lifecycle)
+  end
+
+  defp put_lifecycle(socket, lifecycle) do
+    put_private(socket, @lifecycle, lifecycle)
+  end
+
+  defp put_private(%Socket{private: private} = socket, key, value) when is_atom(key) do
+    %{socket | private: Map.put(private, key, value)}
+  end
+
+  # Lifecycle Event API
+
+  @doc false
+  def mount(params, session, %Socket{private: %{@lifecycle => lifecycle}} = socket) do
+    Enum.reduce(lifecycle.mount, {:ok, socket}, fn %Hook{} = hook, {:ok, socket} ->
+      case hook.function.(params, session, socket) do
+        {:ok, %Socket{}} = state -> state
+        other -> raise_bad_lifecycle_response!(other, socket.view, hook, :mount, 3)
+      end
+    end)
+  end
+
+  @doc false
+  def handle_event(event, val, %Socket{private: %{@lifecycle => lifecycle}} = socket) do
+    Enum.reduce_while(lifecycle.handle_event, {:cont, socket}, fn %Hook{} = hook,
+                                                                  {:cont, socket} ->
+      case hook.function.(event, val, socket) do
+        {:cont, %Socket{}} = state -> {:cont, state}
+        {:halt, %Socket{}} = state -> {:halt, state}
+        other -> raise_bad_lifecycle_response!(other, socket.view, hook, :handle_event, 3)
+      end
+    end)
+  end
+
+  def handle_event(_event, _val, %Socket{} = socket), do: {:cont, socket}
+
+  @doc false
+  def handle_params(params, uri, %Socket{private: %{@lifecycle => hooks}} = socket) do
+    Enum.reduce_while(hooks.handle_params, {:cont, socket}, fn %Hook{} = hook, {:cont, socket} ->
```

Comment:
Perhaps it is worth moving to our own traversal?

    defp reduce_socket([hook | hooks], acc, function) do
      case function.(hook, acc) do
        {:cont, %Socket{} = socket} -> reduce_socket(hooks, socket, function)
        {:halt, %Socket{} = socket} -> {:halt, socket}
        other -> bad_lifecycle!(other, hook)
      end
    end

    defp reduce_socket([], acc, _function), do: {:cont, acc}

The hook has the stage for the error message. :)